ibm/resource_ibm_cloudant.go

```diff
@@ -0,0 +1,603 @@
+	} else {
+		auditEventTypes = []string{"management"}
+	}
+
+	opts := client.NewPostActivityTrackerEventsOptions(auditEventTypes)
+
+	_, response, err := client.PostActivityTrackerEvents(opts)
+	if err != nil {
+		log.Printf("[DEBUG] Error updating activity tracker events: %s\n%s", err, response)
+	}
+	return err
+}
+
+func validateCloudantInstanceCapacity(d *schema.ResourceData) error {
+	plan := d.Get("plan").(string)
+	capacity := d.Get("capacity").(int)
+	if capacity > 1 && plan == "lite" {
+		return fmt.Errorf("Setting capacity is not supported for your instance's plan.")
+	}
+	return nil
+}
+
+func setCloudantInstanceCapacity(client *cloudantv1.CloudantV1, d *schema.ResourceData) error {
+	capacityThroughputInformation, err := readCloudantInstanceCapacity(client)
+	if err != nil {
+		return fmt.Errorf("Error retrieving capacity throughput information: %s", err)
+	}
+
+	if capacityThroughputInformation.Current != nil && capacityThroughputInformation.Current.Throughput != nil {
+		currentThroughput := capacityThroughputInformation.Current.Throughput
+		// lite plan doesn't have "blocks" attr on broker's response
+		if d.Get("plan").(string) == "lite" {
+			d.Set("capacity", 1)
+		} else {
+			blocks := int(*currentThroughput.Blocks)
+			d.Set("capacity", blocks)
+		}
+		throughput := map[string]int{
+			"query": int(*currentThroughput.Query),
+			"read":  int(*currentThroughput.Read),
+			"write": int(*currentThroughput.Write),
+		}
+		d.Set("throughput", throughput)
+	}
+	return nil
+}
+
+func readCloudantInstanceCapacity(client *cloudantv1.CloudantV1) (*cloudantv1.CapacityThroughputInformation, error) {
+	opts := client.NewGetCapacityThroughputInformationOptions()
+
+	capacityThroughputInformation, response, err := client.GetCapacityThroughputInformation(opts)
+	if err != nil {
+		log.Printf("[DEBUG] Error getting capacity throughput information: %s\n%s", err, response)
+	}
+	return capacityThroughputInformation, nil
+}
+
+func updateCloudantInstanceCapacity(client *cloudantv1.CloudantV1, d *schema.ResourceData) error {
+	blocks := int64(d.Get("capacity").(int))
+
+	putOpts := client.NewPutCapacityThroughputConfigurationOptions(blocks)
+
+	_, response, err := client.PutCapacityThroughputConfiguration(putOpts)
+	if err != nil {
+		log.Printf("[DEBUG] Error updating capacity throughput: %s\n%s", err, response)
+		return err
+	}
+
+	return isWaitForCapacityUpdated(client)
+}
+
+func isWaitForCapacityUpdated(client *cloudantv1.CloudantV1) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry"},
+		Target:  []string{"done", "failed"},
+		Refresh: func() (interface{}, string, error) {
+			capacityThroughputInformation, err := readCloudantInstanceCapacity(client)
+			if err != nil {
+				return nil, "failed", err
+			}
+
+			state := "retry"
+			current := *capacityThroughputInformation.Current.Throughput.Blocks
+			target := *capacityThroughputInformation.Target.Throughput.Blocks
+
+			if current == target {
```

Comment:
So based on the conversation we had it seems that

1) we can't use presence of `target` as an indicator of success since it's "set once, stays forever" attribute.

2) while we can pass "proper" target as `int`,  it'll create inconsistent behavior in case in-between updates. Sometimes we'll fail with timeout if `current` just over or jump under passed target and sometime we'll get false positives, e.g. we are changing blocks from 3 to 5, another actor changes blocks from 3 to 8, we see 5 during the update and report success.

So essentially current implementation behaves as the following: It waits for target to match current. Is there is in-between update it'll wait for it, instead of configured value. Then it'll read remote `current` value during "read back after update" and update local capacity state accordingly and get out of sync with config.

I think this behavior is not really fixable if we assume changes to remote state outside of terraform state management and if not then waiting for `target` provides better predictability.

Reply:
Yeah I didn't realise we might see intermediate states too and that muddies the picture even more. Let's stick with it as is.